operator/validator/metadata/sync_batch.go

```diff
@@ -0,0 +1,49 @@
+package metadata
+
+import (
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	spectypes "github.com/ssvlabs/ssv-spec/types"
+
+	ssvtypes "github.com/ssvlabs/ssv/protocol/v2/types"
+)
+
+type SyncBatch struct {
+	SharesBefore []*ssvtypes.SSVShare
+	SharesAfter  []*ssvtypes.SSVShare
+	Epoch        phase0.Epoch
+}
+
+// DetectValidatorStateChanges compares validator metadata before and after the update to detect state changes.
+// Returns validators that became attesting, slashed, or exited.
+func (s SyncBatch) DetectValidatorStateChanges() (attesting, slashed, exited []*ssvtypes.SSVShare) {
+	// Build a map of previous states for quick lookups
+	beforeMap := make(map[spectypes.ValidatorPK]*ssvtypes.SSVShare, len(s.SharesBefore))
+	for _, share := range s.SharesBefore {
+		beforeMap[share.ValidatorPubKey] = share
+	}
+
+	attesting = make([]*ssvtypes.SSVShare, 0, len(s.SharesAfter))
+	slashed = make([]*ssvtypes.SSVShare, 0, len(s.SharesAfter))
+	exited = make([]*ssvtypes.SSVShare, 0, len(s.SharesAfter))
+
+	for _, shareAfter := range s.SharesAfter {
+		shareBefore, exists := beforeMap[shareAfter.ValidatorPubKey]
+		if !exists {
+			continue
+		}
+
+		if !shareBefore.IsAttesting(s.Epoch) && shareAfter.IsAttesting(s.Epoch) {
+			attesting = append(attesting, shareAfter)
+		}
```

Comment:
wouldn't this exclude validators who were pending and now active, so they won't be started?

in these cases the metadata didn't change, but the validator is now active

Reply:
also i don't think it's safe to compare pointers to shares, as updates to the "after" one could modify the "before"